django/db/models/fields/files.py

```diff
@@ -224,6 +225,15 @@ def __init__(self, verbose_name=None, name=None, upload_to='', storage=None, **k
         self._primary_key_set_explicitly = 'primary_key' in kwargs
 
         self.storage = storage or default_storage
+
+        if callable(self.storage):
+            self.storage = self.storage()
```

Comment:
Would it be possible to have the instance passed to the callable? Similar to `upload_to`

Reply:
An instance has always been possible.
I totally forgot to fix this PR, I'll try and do it tonight to get it merged.

Reply:
Fun thing here I just realized.

You can now pass an instance of Storage with an added __call__ method:
```python
class MyStorage(Storage):
    def __call__(self):
        if settings.BLAH:
            self.location = '...'
        return self

my_image = models.ImageField(location=MyStorage())
````